File: bindings/python/insight/signal/radar.py
"""Radar signal processing functions.

Provides pulse compression, pulse-Doppler processing, CFAR detection,
MVDR beamforming, and ambiguity function computation.
"""

from insight._insight import signal as _signal

__all__ = [
    "pulse_compression",
    "pulse_doppler",
    "cfar_alpha",
    "ca_cfar",
    "mvdr",
    "ambgfun",
]


def pulse_compression(x, template_tx, normalize=False, window="", nfft=0):
    """Perform pulse compression using matched filtering.

    Args:
        x: Received signal array.
        template_tx: Transmit template (reference waveform).
        normalize: If True, normalize the output.  Default is False.
        window: Window function name to apply before compression.
            Empty string disables windowing.  Default is ``""``.
        nfft: FFT size.  If 0 (default), uses the next power of two.

    Returns:
        Pulse-compressed output array.
    """
    return _signal.pulse_compression(x, template_tx, normalize=normalize,
                                     window=window, nfft=nfft)


def pulse_doppler(x, window="", nfft=0):
    """Compute the pulse-Doppler map from a slow-time matrix.

    Args:
        x: Input 2-D array of shape ``(num_pulses, num_range_bins)``.
        window: Window function name to apply along the pulse axis.
            Empty string disables windowing.  Default is ``""``.
        nfft: FFT size for the Doppler dimension.  If 0 (default), uses
            the number of pulses.

    Returns:
        2-D pulse-Doppler map array.
    """
    return _signal.pulse_doppler(x, window=window, nfft=nfft)


def cfar_alpha(pfa, N):
    """Compute the CFAR threshold multiplier for a given PFA and window size.

    Args:
        pfa: Probability of false alarm.
        N: Number of reference cells.

    Returns:
        CFAR scaling factor (alpha).
    """
    return _signal.cfar_alpha(pfa, N)


def ca_cfar(data, guard_cells, reference_cells, pfa=1e-3):
    """Apply Cell-Averaging CFAR (CA-CFAR) detection.

    Args:
        data: Input array (1-D or N-D).
        guard_cells: Number of guard cells on each side per dimension.
        reference_cells: Number of reference cells on each side per
            dimension.
        pfa: Probability of false alarm.  Default is 1e-3.

    Returns:
        Boolean mask array of detected targets (same shape as *data*).
    """
    return _signal.ca_cfar(data, list(guard_cells), list(reference_cells),
                           pfa=pfa)


def mvdr(x, sv, calc_cov=True):
    """Apply MVDR (Capon) beamforming.

    Args:
        x: Input signal matrix of shape ``(num_elements, num_snapshots)``.
        sv: Steering vector.
        calc_cov: If True (default), estimate the covariance matrix from
            *x*.  If False, *x* is used as the covariance matrix directly.

    Returns:
        Beamformed output array.
    """
    return _signal.mvdr(x, sv, calc_cov=calc_cov)


def ambgfun(x, fs, prf=1.0, y=None, cut="2d", cutValue=0.0):
    """Compute the ambiguity function.

    Args:
        x: Input signal array.
        fs: Sampling frequency in Hz.
        prf: Pulse repetition frequency in Hz.  Default is 1.0.
        y: Optional second signal for cross-ambiguity.  If ``None``,
            computes the auto-ambiguity.
        cut: Cut type.  ``'2d'`` (default) returns the full 2-D ambiguity
            surface; ``'delay'`` returns a cut at constant Doppler;
            ``'doppler'`` returns a cut at constant delay.
        cutValue: Value at which to take the cut (in seconds for delay
            cut, in Hz for Doppler cut).  Default is 0.0.

    Returns:
        Ambiguity function array (1-D or 2-D depending on *cut*).
    """
    kwargs = {}
    if y is not None:
        kwargs["y"] = y
    return _signal.ambgfun(x, fs, prf, cut=cut, cutValue=cutValue, **kwargs)

File: bindings/lua/insight/signal/radar.lua
--- Radar signal processing functions.
-- Provides pulse compression, Doppler processing, CFAR detection,
-- MVDR beamforming, and ambiguity function computation.
-- @module insight.signal.radar

local native = require("_insight")
local sig = native.signal

local M = {}

local _wrap = require("insight._wrap")

--- Perform pulse compression using matched filtering.
-- @tparam Array signal Input radar signal (1-D or 2-D).
-- @tparam Array reference Reference waveform (1-D).
-- @treturn Array Compressed pulse.
M.pulse_compression = _wrap({ "signal", "reference" }, function(signal, reference)
  return sig.pulse_compression(signal, reference)
end)

--- Perform pulse-Doppler processing.
-- @tparam Array data Input CPI data (2-D: pulses x range bins).
-- @string[opt=""] window Window name (e.g., "hann"). Empty for none.
-- @int[opt=0] nfft Number of FFT points for Doppler processing (0 = num_pulses).
-- @treturn Array Doppler-range map.
M.pulse_doppler = _wrap({ "data", "window", "nfft" }, function(data, window, nfft)
  return sig.pulse_doppler(data, window or "", nfft or 0)
end)

--- Compute CFAR threshold alpha coefficient.
-- @number pfa Probability of false alarm.
-- @int N Number of reference cells.
-- @treturn number CFAR alpha threshold coefficient.
M.cfar_alpha = _wrap({ "pfa", "N" }, function(pfa, N)
  return sig.cfar_alpha(pfa, N)
end)

--- Apply Cell-Averaging CFAR detection.
-- @tparam Array data Input signal array.
-- @table guard_cells Guard cells per dimension, e.g. {2} or {2, 2}.
-- @table ref_cells Reference cells per dimension, e.g. {3} or {3, 3}.
-- @number[opt=1e-3] pfa Probability of false alarm.
-- @treturn table Pair of {threshold, detections} arrays.
M.ca_cfar = _wrap({ "data", "guard_cells", "ref_cells", "pfa" }, function(data, guard_cells, ref_cells, pfa)
  return sig.ca_cfar(data, guard_cells, ref_cells, pfa or 1e-3)
end)

--- Perform MVDR (Capon) beamforming.
-- @tparam Array x Input data matrix (sensors x samples).
-- @tparam Array steering Steering vector.
-- @treturn Array MVDR beamformer output.
M.mvdr = _wrap({ "x", "steering" }, function(x, steering)
  return sig.mvdr(x, steering)
end)

--- Compute the ambiguity function.
-- @tparam Array x Input signal (1-D, complex).
-- @number fs Sampling frequency in Hz.
-- @number prf Pulse repetition frequency in Hz.
-- @treturn Array Ambiguity function (2-D: delay x Doppler).
M.ambgfun = _wrap({ "x", "fs", "prf" }, function(x, fs, prf)
  return sig.ambgfun(x, fs, prf)
end)

return M
